ethers-providers/src/provider.rs

```diff
@@ -786,6 +786,151 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
+                            data: Some([&erc::ERC721_SELECTOR[..], &token_id].concat().into()),
+                            to: Some(NameOrAddress::Address(contract_addr)),
+                            ..Default::default()
+                        };
+                        let data = self.call(&tx.into(), None).await?;
+                        if decode_bytes::<Address>(ParamType::Address, data) != owner {
+                            return Err(ProviderError::CustomError("Incorrect owner.".to_string()));
+                        }
+                        [0xc8, 0x7b, 0x56, 0xdd]
+                    }
+                    "erc1155" => {
+                        let tx = TransactionRequest {
+                            data: Some(
+                                [&erc::ERC1155_SELECTOR[..], &[0x0; 12], &owner.0, &token_id]
+                                    .concat()
+                                    .into(),
+                            ),
+                            to: Some(NameOrAddress::Address(contract_addr)),
+                            ..Default::default()
+                        };
+                        let data = self.call(&tx.into(), None).await?;
+                        if decode_bytes::<u64>(ParamType::Uint(64), data) == 0 {
+                            return Err(ProviderError::CustomError(
+                                "Incorrect balance.".to_string(),
+                            ));
+                        }
+                        [0x0e, 0x89, 0x34, 0x1c]
+                    }
+                    _ => {
+                        return Err(ProviderError::CustomError(
+                            "Unsupported eip155 token type".to_string(),
+                        ))
+                    }
+                };
+
+                let tx = TransactionRequest {
+                    data: Some([&selector[..], &token_id].concat().into()),
+                    to: Some(NameOrAddress::Address(contract_addr)),
+                    ..Default::default()
+                };
+                let data = self.call(&tx.into(), None).await?;
+                let mut metadata_url = Url::parse(&decode_bytes::<String>(ParamType::String, data))
+                    .map_err(|e| {
+                        ProviderError::CustomError(format!("Invalid metadata url: {}", e))
+                    })?;
+
+                if inner_scheme == "erc1155" {
+                    metadata_url.set_path(
+                        &metadata_url.path().replace("%7Bid%7D", &hex::encode(&token_id)),
+                    );
+                }
+                if metadata_url.scheme() == "ipfs" {
+                    metadata_url =
+                        erc::http_link_ipfs(metadata_url).map_err(ProviderError::CustomError)?;
+                }
+                let metadata: erc::Metadata = reqwest::get(metadata_url)
+                    .await
+                    .map_err(|e| ProviderError::CustomError(e.to_string()))?
```

Comment:
is this necessary?
does ProviderError not have a variant that represents the reqwest Error type? if not probably worth adding so we can use `?` instead

Reply:
Improved in 78d6118